### ISSUE TITLE
Expandable Memory-Mapped File Buffers

### ISSUE THREAD

**Issue by abyrd** [RuedigerMoeller/fast-serialization#175]
It has been established in #132 that FSTStreamEncoder cannot currently flush its internal buffer to the output stream when it gets full. The buffer is flushed to the stream only after each top-level object is completely serialized, including its tree of referenced objects. Therefore the buffer must grow.

This is not a problem when serializing small objects, but one use case for FST is persistence of large data structures for re-use when a program restarts. We frequently serialize and re-load object trees in the 2GB-3GB range in our projects. In this case, serialization can double the memory consumption. Both the object tree and its serialized form must exist in memory at the same time, and some objects like primitive arrays are about the same size in memory or serialized. 

An alternative to serializing to an output stream is serializing to a file that has been mapped to a memory address range and letting the operating system worry about flushing to disk and swapping pages in and out. However, we have no idea how large the final file will be, so there must be some mechanism for truncating the file and ideally a mechanism for growing the mapped memory region while it's being written. Otherwise we must guess the maximum size of the file or allocate Integer.MAX_VALUE which works but is kind of ridiculous.

I now have two implementations of this idea, one based on 
a ByteBufferBasicBytez wrapping a buffer created by FileChannel.map(), and the other based on unsafe access to memory mapped with the native map and unmap methods of a FileChannel. The former only allows 31-bit addressing (2GB max file) and the latter in theory allows 63-bit addressing (though some operating systems will limit you to around 4GB).

**Comment by abyrd** [comment]
@RuedigerMoeller I can submit a pull request with these two implementations as soon as #173 and #174 are fixed.

**Comment by RuedigerMoeller** [comment]
hi, i've fixed #173, #174. Have you seen MMFBytez class ? It's a memory mapped file backed implementation of bytesource.
Problem with MMF is that upon resizing one must unmap the file using pricate api, then append the number of bytes to the underlying file (slowish) and remap it.

**Comment by abyrd** [comment]
@RuedigerMoeller I've got an example class already prepared, should have a PR within a few days.

**Comment by RuedigerMoeller** [comment]
You have seen that memory mapped buffer is already present (to avoid duplication) ?

**Comment by abyrd** [comment]
Yes, I experimented with MMFBytez and its design before making my own. There are comments in my code that explain the difference. If I remember correctly the existing MMFBytez required knowing the size of the buffer before serialization. Mine expands linearly whenever it does not have enough space (but does so by interpreting the requested expanded buffer size as a hint rather than a requirement). It also uses the unsafe methods to allow 63-bit addressing. It won't be hard to contribute the code, then we can see what you think of it.

**Comment by abyrd** [comment]
Really though the only reason I ended up using memory mapped files was that I'm serializing huge objects and current design can only serialize to a single contiguous buffer. So I see #132 as the core issue.

**Comment by RuedigerMoeller** [comment]
ah .. Streaming serialization probably won't be implemented anytime soon as it also imposes checks and arithmetics with every byte written/read (can have significant impact as its a core operation). As memory has become cheap the need for this has also decreased (except some edge cases). In addition large plain byte array buffers do not hit GC duration too much (its actually the number of objects which determine GC overhead), so .. well .. ;)
